### PR TITLE
Add "the Rustc book"

### PR DESCRIPTION
**Pull request by steveklabnik**
This PR introduces a new book into the documentation, "The rustc book". We already have books for Cargo, and for Rustdoc, rustc should have some too. This book is focused on *users* of rustc, and provides a nice place to write documentation for users.

I haven't put content here, but plan on scaffolding it out very soon, and wanted this PR open for a few discussions first. One of those is "what exactly should said TOC be?" I plan on having a proposed one up tomorrow, but figured I'd let people know to start thinking about it now.

The big one is that we also will want to put https://github.com/rust-lang-nursery/rustc-guide in-tree as well, and the naming is... tough. I'm proposing:

* doc.rust-lang.org/rustc is "The Rustc book", to mirror the other tools' books.
* doc.rust-lang.org/rustc-contribution is "The Rustc contribution guide", and contains that book

@nikomatsakis et al, any thoughts on this? I'm not attached to it in particular, but had to put something together to get this discussion going. I think mirroring the other tools is a good idea for this work, but am not sure where exactly that leaves yours.

Fixes https://github.com/rust-docs/team/issues/11

### CODE REVIEW

**Comment by zackmdavis** [comment]
I wouldn't have expected there to be enough material to cover in a special rustc-the-compiler book that doesn't already belong in the [Rust-the-language book](https://doc.rust-lang.org/book/second-edition/)?

**Comment by mark-i-m** [comment]
> - doc.rust-lang.org/rustc is "The Rustc book", to mirror the other tools' books.
> - doc.rust-lang.org/rustc-contribution is "The Rustc contribution guide", and contains that book

How about
- d.rl.o/rustc-user is "The Rustc User Guide"
- d.rl.o/rustc-contrib is "The Rustc Contributer Guide"

**Comment by mark-i-m** [comment]
> The big one is that we also will want to put https://github.com/rust-lang-nursery/rustc-guide in-tree

Also, do you have a time line for when you want to do that? (cc #48478)

**Comment by steveklabnik** [comment]
> Also, do you have a time line for when you want to do that? 

That's more up to them than to me, whenever they're comfy doing so, I'm happy to make it happen!

**Comment by TimNN** [comment]
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/363194154). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Receiving objects:  83% (633653/754749), 308.45 MiB | 1024.00 KiB/s
Receiving objects:  84% (633990/754749), 308.45 MiB | 1024.00 KiB/s
---
Resolving deltas: 100% (612506/612506), completed with 4864 local objects.
---
[00:00:52] configure: rust.quiet-tests     := True
---
[00:58:37] ..........................................................................i.........................
[00:58:45] .................i..................................................................................
---
[00:59:29] ..............................................................................................i.....
[00:59:38] ....................................................................i...............................
---
[01:00:54] .............................................i......................................................
---
[01:06:15] .............................i......................................................................
[01:06:34] ..............................................................i.....................................
[01:06:54] ...............................................i....................................................
[01:07:20] ....................................................................................................
[01:07:48] ....................................................................................................
[01:08:16] ....................................................................................................
[01:08:49] ..i...............................................................................................i.
[01:09:06] ......................................test [run-pass] run-pass/mir_heavy_promoted.rs has been running for over 60 seconds
[01:09:38] ..............................................................
[01:10:18] ....................................................................................................
[01:11:06] ..............................................................ii....................................
[01:11:45] .........................i.........................test [run-pass] run-pass/saturating-float-casts.rs has been running for over 60 seconds
[01:12:11] ...........................i.ii..................
[01:13:04] ......................................................................................iiiiiii.......
---
[01:16:30] ..................i............................................................ii.iii...............
[01:16:40] ....................................................................................................
[01:16:50] ........i..............................i............................................................
[01:16:59] ....................................................................................................
[01:17:08] ....................i...............................................................................
---
[01:18:24] .............i......................................................................................
[01:18:37] .................i..ii..............................................................................
[01:18:49] ....................................................................................................
[01:19:00] ....................................................................................................
[01:19:11] ....................................................................................i...............
[01:19:23] ..............................i.....................................................................
---
[01:20:06] ...........................i........................................................................
[01:20:08] ....................................................................i...............................
[01:20:09] ................i.......................................................
---
[01:20:29] ...........i........................
---
[01:21:05] i...i..ii....i.............ii........iii......i..i...i...ii..i..i..ii.....
---
[01:21:09] i.......i......................i......
---
[01:22:00] iiii.......i..i........i..i.i.............i..........iiii...........i...i..........ii.i.i.......ii..
[01:22:01] ....ii...
---
[01:35:13] .....i..............................................................................................
---
[01:38:10] .....................................i..............................................................
[01:38:41] ....................................................................................................
[01:39:12] ..........................................i.........................................................
---
[01:41:21] .........................................................ii.........................................
---
[01:42:50] ..............................................................i.....................................
---
[01:49:27] ii..................................................................................................
[01:49:50] ....................................................................................................
[01:50:10] ...................iii......i......i...i......i.....................................................
[01:50:23] ....................................................................................................
[01:50:43] .........................................iiii........ii.............................................
[01:50:57] ....................................................................................................
[01:51:18] ........................................................................................i...........
[01:51:51] ....................................................................................................
[01:52:05] ..................................................................................................ii
[01:52:18] ii...............................................
---
[02:05:13] command did not execute successfully: "/checkout/obj/build/bootstrap/debug/rustdoc" "--test" "/checkout/src/doc/rustc/src/lints/index.md" "--test-args" ""
---
[02:05:13] ---- /checkout/src/doc/rustc/src/lints/index.md - Lints (line 10) stdout ----
[02:05:13]  error: unknown start of token: `
[02:05:13]  --> /checkout/src/doc/rustc/src/lints/index.md:16:27
[02:05:13]   |
[02:05:13] 7 | warning: unused variable: `x`
[02:05:13]   |                           ^
[02:05:13]   |
[02:05:13] help: unicode character '`' (Grave Accent) looks like ''' (Single Quote), but it's not
[02:05:13]  --> /checkout/src/doc/rustc/src/lints/index.md:16:27
[02:05:13]   |
[02:05:13] 7 | warning: unused variable: `x`
[02:05:13]   |                           ^
[02:05:13]
[02:05:13] thread '/checkout/src/doc/rustc/src/lints/index.md - Lints (line 10)' panicked at 'couldn't compile the test', librustdoc/test.rs:306:13
---
[02:05:13] Makefile:58: recipe for target 'check' failed
[02:05:13] make: *** [check] Error 1
---
$ ls -lat $HOME/Library/Logs/DiagnosticReports/
ls: cannot access /home/travis/Library/Logs/DiagnosticReports/: No such file or directory
```

</details><p></p>

I'm a bot! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**.

**Comment by TimNN** [comment]
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/363209526). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Resolving deltas: 100% (612513/612513), completed with 4864 local objects.
---
[00:00:45] configure: rust.quiet-tests     := True
---
[00:43:55] ..........................................................................i.........................
[00:44:01] .................i..................................................................................
---
[00:44:37] ..............................................................................................i.....
[00:44:44] ....................................................................i...............................
---
[00:45:41] .............................................i......................................................
---
[00:49:42] .............................i......................................................................
[00:49:56] ..............................................................i.....................................
[00:50:12] ...............................................i....................................................
[00:50:33] ....................................................................................................
[00:50:55] ....................................................................................................
[00:51:17] ....................................................................................................
[00:51:43] ..i...............................................................................................i.
[00:52:09] .............................................................................test [run-pass] run-pass/mir_heavy_promoted.rs has been running for over 60 seconds
[00:52:21] .......................
[00:52:52] ....................................................................................................
[00:53:30] ..............................................................ii....................................
[00:54:14] .........................i....................................................itest [run-pass] run-pass/saturating-float-casts.rs has been running for over 60 seconds
[00:54:23] .ii..................
[00:55:06] ......................................................................................iiiiiii.......
---
[00:57:20] ..................i............................................................ii.iii...............
[00:57:28] ....................................................................................................
[00:57:35] ........i..............................i............................................................
[00:57:43] ....................................................................................................
[00:57:50] ....................i...............................................................................
[00:57:58] ....................................................................................................
[00:58:08] ....................................................................................................
[00:58:19] ....................................................................................................
[00:58:29] ....................................................................................................
[00:58:43] ....................................................................................................
[00:58:53] .............i......................................................................................
[00:59:02] .................i..ii..............................................................................
[00:59:12] ....................................................................................................
[00:59:22] ....................................................................................................
[00:59:32] ....................................................................................i...............
[00:59:42] ..............................i.....................................................................
---
[01:00:19] ...........................i........................................................................
[01:00:21] ....................................................................i...............................
[01:00:22] ................i.......................................................
---
[01:00:37] ...........i........................
---
[01:01:07] i...i..ii....i.............ii........iii......i..i...i...ii..i..i..ii.....
---
[01:01:10] i.......i......................i......
---
[01:01:49] iiii.......i..i........i..i.i.............i..........iiii...........i...i..........ii.i.i.......ii..
[01:01:50] ....ii...
---
[01:11:37] .....i..............................................................................................
---
[01:13:38] .....................................i..............................................................
[01:13:59] ....................................................................................................
[01:14:21] ..........................................i.........................................................
---
[01:15:50] .........................................................ii.........................................
---
[01:16:56] ..............................................................i.....................................
---
[01:21:57] ii..................................................................................................
[01:22:16] ....................................................................................................
[01:22:32] ...................iii......i......i...i......i.....................................................
[01:22:41] ....................................................................................................
[01:22:57] .........................................iiii........ii.............................................
[01:23:08] ....................................................................................................
[01:23:24] ........................................................................................i...........
[01:23:49] ....................................................................................................
[01:24:01] ..................................................................................................ii
[01:24:10] ii...............................................
---
[01:34:33] command did not execute successfully: "/checkout/obj/build/bootstrap/debug/rustdoc" "--test" "/checkout/src/doc/rustc/src/lints/index.md" "--test-args" ""
---
[01:34:33] ---- /checkout/src/doc/rustc/src/lints/index.md - Lints (line 10) stdout ----
[01:34:33]  error: unknown start of token: `
[01:34:33]  --> /checkout/src/doc/rustc/src/lints/index.md:16:27
[01:34:33]   |
[01:34:33] 7 | warning: unused variable: `x`
[01:34:33]   |                           ^
[01:34:33]   |
[01:34:33] help: unicode character '`' (Grave Accent) looks like ''' (Single Quote), but it's not
[01:34:33]  --> /checkout/src/doc/rustc/src/lints/index.md:16:27
[01:34:33]   |
[01:34:33] 7 | warning: unused variable: `x`
[01:34:33]   |                           ^
[01:34:33]
[01:34:33] thread '/checkout/src/doc/rustc/src/lints/index.md - Lints (line 10)' panicked at 'couldn't compile the test', librustdoc/test.rs:306:13
---
[01:34:33] Makefile:58: recipe for target 'check' failed
[01:34:33] make: *** [check] Error 1
---
$ dmesg | grep -i kill
[   10.793162] init: failsafe main process (1094) killed by TERM signal
```

</details><p></p>

I'm a bot! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**.

File: src/doc/rustc/src/command-line-arguments.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+## `-v`/`--verbose`: use verbose output
+
+## `--extern`: specify where an external library is located
+
+## `--sysroot`: Override the system root
+
+## `--error-format`: control how errors are produced
+
+## `--color`: configur coloring of output
```
Comment (goodmanjonathan) [inline src/doc/rustc/src/command-line-arguments.md]: typo: configure

**Comment by TimNN** [comment]
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/363258957). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Resolving deltas: 100% (612544/612544), completed with 4858 local objects.
---
[00:00:53] configure: rust.quiet-tests     := True
---
[00:38:18] ..........................................................................i.........................
[00:38:24] .................i..................................................................................
---
[00:38:57] ..............................................................................................i.....
[00:39:04] ....................................................................i...............................
---
[00:39:56] .............................................i......................................................
---
[00:43:37] .............................i......................................................................
[00:43:51] ..............................................................i.....................................
[00:44:06] ...............................................i....................................................
[00:44:25] ....................................................................................................
[00:44:45] ....................................................................................................
[00:45:05] ....................................................................................................
[00:45:29] ..i...............................................................................................i.
[00:45:58] ..........................................................................................test [run-pass] run-pass/mir_heavy_promoted.rs has been running for over 60 seconds
[00:46:03] ..........
[00:46:32] ....................................................................................................
[00:47:05] ..............................................................ii....................................
[00:47:52] .........................i....................................................i.ii...............test [run-pass] run-pass/saturating-float-casts.rs has been running for over 60 seconds
[00:47:53] ...
[00:48:30] ......................................................................................iiiiiii.......
---
[00:50:39] ..................i............................................................ii.iii...............
[00:50:47] ....................................................................................................
[00:50:54] ........i..............................i............................................................
[00:51:01] ....................................................................................................
[00:51:08] ....................i...............................................................................
[00:51:16] ....................................................................................................
[00:51:25] ....................................................................................................
[00:51:35] ....................................................................................................
[00:51:46] ....................................................................................................
[00:51:59] ....................................................................................................
[00:52:07] .............i......................................................................................
[00:52:16] .................i..ii..............................................................................
[00:52:26] ....................................................................................................
[00:52:35] ....................................................................................................
[00:52:45] ....................................................................................i...............
[00:52:55] ..............................i.....................................................................
---
[00:53:30] ...........................i........................................................................
[00:53:32] ....................................................................i...............................
[00:53:33] ................i.......................................................
---
[00:53:47] ...........i........................
---
[00:54:15] i...i..ii....i.............ii.........iii......i..i...i...ii..i..i..ii.....
---
[00:54:18] i.......i......................i.......
---
[00:54:54] iiii.......i..i........i..i.i.............i..........iiii...........i...i..........ii.i.i.......ii..
[00:54:55] ....ii...
---
[01:03:07] .....i..............................................................................................
---
[01:04:51] .....................................i..............................................................
[01:05:08] ....................................................................................................
[01:05:26] ..........................................i.........................................................
---
[01:06:45] .........................................................ii.........................................
---
[01:07:42] ..............................................................i.....................................
---
[01:12:03] ii..................................................................................................
[01:12:20] ....................................................................................................
[01:12:34] ...................iii......i......i...i......i.....................................................
[01:12:42] ....................................................................................................
[01:12:56] .........................................iiii........ii.............................................
[01:13:05] ....................................................................................................
[01:13:20] ........................................................................................i...........
[01:13:41] ....................................................................................................
[01:13:51] ..................................................................................................ii
[01:13:59] ii...............................................
---
[01:23:10] command did not execute successfully: "/checkout/obj/build/bootstrap/debug/rustdoc" "--test" "/checkout/src/doc/rustc/src/lints/index.md" "--test-args" ""
---
[01:23:10] ---- /checkout/src/doc/rustc/src/lints/index.md - Lints (line 10) stdout ----
[01:23:10]  error: unknown start of token: `
[01:23:10]  --> /checkout/src/doc/rustc/src/lints/index.md:16:27
[01:23:10]   |
[01:23:10] 7 | warning: unused variable: `x`
[01:23:10]   |                           ^
[01:23:10]   |
[01:23:10] help: unicode character '`' (Grave Accent) looks like ''' (Single Quote), but it's not
[01:23:10]  --> /checkout/src/doc/rustc/src/lints/index.md:16:27
[01:23:10]   |
[01:23:10] 7 | warning: unused variable: `x`
[01:23:10]   |                           ^
[01:23:10]
[01:23:10] thread '/checkout/src/doc/rustc/src/lints/index.md - Lints (line 10)' panicked at 'couldn't compile the test', librustdoc/test.rs:306:13
---
[01:23:10] Makefile:58: recipe for target 'check' failed
[01:23:10] make: *** [check] Error 1
---
$ ls -lat $HOME/Library/Logs/DiagnosticReports/
ls: cannot access /home/travis/Library/Logs/DiagnosticReports/: No such file or directory
travis_time:end:166f3118:start=1523047194621800172,finish=1523047194635674619,duration=13874447
travis_fold:end:after_failure.2
travis_fold:start:after_failure.3
travis_time:start:09a45ee4
$ find $HOME/Library/Logs/DiagnosticReports -type f -name '*.crash' -not -name '*.stage2-*.crash' -not -name 'com.apple.CoreSimulator.CoreSimulatorService-*.crash' -exec printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" {} \; -exec head -750 {} \; -exec echo travis_fold":"end:crashlog \; || true
find: `/home/travis/Library/Logs/DiagnosticReports': No such file or directory
travis_time:end:09a45ee4:start=1523047194641609724,finish=1523047194647690626,duration=6080902
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0aff7134
$ dmesg | grep -i kill
[   10.320157] init: failsafe main process (1093) killed by TERM signal
```

</details><p></p>

I'm a bot! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**.

**Comment by phansch** [comment]
I believe the rustc lint documentation would partly resolve https://github.com/rust-lang/rust/issues/49416 as well :+1:

**Comment by nikomatsakis** [comment]
I think that the naming of "Rustc contributor guide" feels pretty ok. It doesn't sound quite as "weighy" as it might -- like it sounds sort of like "CONTRIBUTING.md" to me -- but then again once you open it, you'll see what it is.  And certainly it contains material that would ordinarily be covered by a typical CONTRIBUTING.md -- that is, I envision it being the "one stop shop" for people who want to hack on rustc.

cc @mark-i-m, who may have an opinion here

**Comment by nikomatsakis** [comment]
Oh I see @mark-i-m already commented :) d'oh

**Comment by nikomatsakis** [comment]
I think "hacker" is often considered to be a kind of exclusionary term.

**Comment by pietroalbini** [comment]
Ping from triage! What's the status of this PR? Unfortunately we don't have a `S-waiting-on-bikeshed` label...

**Comment by mark-i-m** [comment]
@pietroalbini Could _please_ create one!?

**Comment by steveklabnik** [comment]
@pietroalbini status is, i am almost ready to :shipit: , should be today.

**Comment by pietroalbini** [comment]
@mark-i-m Should I really do that? I thought the plan was to merge PRs in the future...

**Comment by mark-i-m** [comment]
@pietroalbini I was half-joking :stuck_out_tongue: ... I would really like to see `waiting-on-bikeshed` in the github issue tracker somewhere :)

**Comment by steveklabnik** [comment]
Okay! This is still a bit bare-bones in many places, but the idea was just to get the scaffolding going. Given the discussion about names above, this also seems fine to ship; we can re-bikeshed the contribution guide later, but it seems like there's no objection to using `/rustc` for this.

File: src/doc/rustc/src/targets/index.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# Targets
+
+`rustc` is a cross-compiler by default. This means that you can use any compiler to build for any
+architecture. The list of *targets* are the possible architectures that you can build for.
+
+You can see the API docs for a given target
+[here](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_back/target/struct.Target.html), all
```
Comment (QuietMisdreavus) [inline src/doc/rustc/src/targets/index.md]: This line seems a little off. My reading implies "this is the standard library available under any given target", but it's more like "these are the things that a target can specify".

**Review by QuietMisdreavus — COMMENTED** [review]
Looks good! Just a few small nits.

File: src/doc/rustc/src/codegen-options/index.md [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+## no-vectorize-loops
+
+By default, `rustc` will attempt to [vectorize
+loops](https://llvm.org/docs/Vectorizers.html#the-loop-vectorizer). This
+flag will turn that behavior off.
+
+## no-vectorize-slp
+
+By default, `rustc` will attempt to vectorize loops using [superword-level
+parallism](https://llvm.org/docs/Vectorizers.html#the-slp-vectorizer). This
```
Comment (sanxiyn) [inline src/doc/rustc/src/codegen-options/index.md]: parallelism

**Comment by illustrious-you** [comment]
How about "The Rustc Book" (doc.rust-lang.org/rustc) and "The Rustc Internals Book" (doc.rust-lang.org/rustc-internals)?

**Comment by mark-i-m** [comment]
> "The Rustc Internals Book" (doc.rust-lang.org/rustc-internals)?

"Internals" doesn't really strike me as "how rustc works" somehow... It's not really what I would google if I was looking for it.

I've also thought about longer titles, like "How rustc works and how to contribute to it"

**Comment by steveklabnik** [comment]
Updated to fix nits.

@illustrious-you i like it!

**Comment by illustrious-you** [comment]
@mark-i-m 

Official documentation from other open-source projects often refers to "internals":

* [GCC](https://gcc.gnu.org/onlinedocs/gccint/)
* [Clang](https://clang.llvm.org/docs/InternalsManual.html)
* [PostgreSQL](https://www.postgresql.org/docs/9.1/static/internals.html)
* [MySQL](https://dev.mysql.com/doc/internals/en/)
* [MariaDB](https://mariadb.com/kb/en/library/mariadb-internals-documentation/)
* [Git](https://git-scm.com/book/en/v2/Git-Internals-Plumbing-and-Porcelain)

While there is no standard for this kind of documentation, there is a norm.

> "Internals" doesn't really strike me as "how rustc works" somehow... It's not really what I would google if I was looking for it.

Huh. When I'm looking for a conceptual overview, it's my go-to term. I'm not sure where I picked that habit up, but it tends to produce reliable results*:

* C- [Google](https://www.google.com/search?q=C+internals), [Bing](https://www.bing.com/search?q=C+internals)
* C++ - [Google](https://www.google.com/search?q=C%2B%2B+internals), [Bing](https://www.bing.com/search?q=C%2B%2B+internals)
* JavaScript - [Google](https://www.google.com/search?q=javascript+internals), [Bing](https://www.bing.com/search?q=Javascript+internals)
* Python - [Google](https://www.google.com/search?q=python+internals), [Bing](https://www.bing.com/search?q=python+internals)
* .Net - [Google](https://www.google.com/search?q=.Net+internals), [Bing](https://www.bing.com/search?q=.Net+internals)
* Go - [Google](https://www.google.com/search?q=go+internals), [Bing](https://www.bing.com/search?q=go+internals)

It's notable, however, that you said it's not the first thing you'd search for Rust, and the search results seem to bear that out: [Google](https://www.google.com/search?q=rust+internals), [Bing](https://www.bing.com/search?q=rust+internals). Instead of finding articles about Rust internals, the search engines find the internals boards.

* Note: to avoid biasing results with your search history, open the links in a private browser session.

**Comment by QuietMisdreavus** [comment]
Unless anyone has any major follow up issues, i'm ready to call this good! Little things can go in as PRs after this lands.

@bors r+

**Comment by pietroalbini** [comment]
Spurious failure compiling `rustc_trans`.

@bors: retry

**Pull request merged**